controllers/virtualmachinesnapshot/virtualmachinesnapshot_controller.go

```diff
@@ -158,3 +184,213 @@ func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineSnapshotContext)
 	ctx.Logger.Info("Successfully patched VirtualMachine's current snapshot reference", "vm.Name", vm.Name, "spec.currentSnapshot", vm.Spec.CurrentSnapshot.Name)
 	return nil
 }
+
+func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VirtualMachineSnapshotContext) error {
+	ctx.Logger.Info("Reconciling VirtualMachineSnapshot deletion")
+	vmSnapshot := ctx.VirtualMachineSnapshot
+	if controllerutil.ContainsFinalizer(vmSnapshot, Finalizer) {
+		vm := &vmopv1.VirtualMachine{}
+		objKey := client.ObjectKey{Name: vmSnapshot.Spec.VMRef.Name, Namespace: vmSnapshot.Namespace}
+		if err := r.Get(ctx, objKey, vm); err != nil {
+			if apierrors.IsNotFound(err) {
+				ctx.Logger.Info("VirtualMachine not found, assuming the snapshot is deleted along with moVM, remove finalizer")
+				controllerutil.RemoveFinalizer(vmSnapshot, Finalizer)
+				return nil
+			}
+			ctx.Logger.Error(err, "failed to get VirtualMachine", "vm", objKey)
+			return err
+		}
+
+		ctx.VM = vm
+		if err := r.deleteSnapshot(ctx); err != nil {
+			if errors.Is(err, vcenter.ErrVMNotFound) { //nolint:gocritic
```

Comment:
If we're check these underlying errors IMO they should be a part of the provider interface instead of leaking the underlying packages here. It might be easier to encode as a additional bool return value instead.

Reply:
Then we would need 2 different bool, one for indicating VM is gone or the snapshot is gone. Would that seems too much? I can move those predefined error to vmprovider interface level as well.